osu.Desktop/Windows/CompatibilityModeChecker.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Versioning;
+using Microsoft.Win32;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Graphics;
+using osu.Game.Overlays;
+using osu.Game.Overlays.Notifications;
+
+namespace osu.Desktop.Windows
+{
+    /// <summary>
+    /// Checks if the game is running with windows compatibility optimizations which could cause issues. Displays a warning notification if so.
+    /// </summary>
+    [SupportedOSPlatform("windows")]
+    public partial class CompatibilityModeChecker : Component
+    {
+        [Resolved]
+        private INotificationOverlay notifications { get; set; } = null!;
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            if (CheckCompatibilityMode())
+                notifications.Post(new CompatibilityModeNotification());
+        }
+
+        public static bool CheckCompatibilityMode()
+        {
+            using var layers = Registry.CurrentUser.OpenSubKey(@"Software\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers");
+
+            string exePath = Assembly.GetExecutingAssembly().Location;
```

Comment:
This doesn't actually work. Contrary to the variable name, this returns the full path of `osu!.dll`, but the registry has the path of `osu!.exe`.

There already exists a perfectly good way to get the exe path in `WindowsAssociationManager`. You should move that to a common windows helper class and use it instead.

https://github.com/ppy/osu/blob/3df32638c2235029625d929c41ce6237c646e75c/osu.Desktop/Windows/WindowsAssociationManager.cs#L32

---

It's also important to note that enabling compatibility mode on the osu!(lazer) start menu shortcut enables it only on the squirrel shim `osu!.exe` and not `osu!.exe` that is tested here. So I'm not certain that this will actually work on deployed builds.

Reply:
[maybe scroll up a few](https://github.com/ppy/osu/pull/27654#issuecomment-2013969557)

Reply:
The logic for locating the executable was incorrect in the initial PR. @bdach gave a patch earlier, replacing `Assembly.GetExecutingAssembly().Location;` with `Environment.ProcessPath` which fixes this issue. I'm going to commit this patch along with some other final touches.

Unless `Environment.ProcessPath` doesn't work (which it should, since I've tested it on deployed builds), we should be fine with using that.